prez/services/query_generation/cql_functions.py

```diff
@@ -0,0 +1,203 @@
+    convert_value_to_rdf_term,
+    create_values_constraint,
+    create_tssp_alt_or_alt_inverse, create_tssp_sequence, create_union_gpnt_from_tssps,
+)
+
+REGISTERED_CQL_FUNCTIONS = [
+    "FOIObservationFilterDirect",
+    "FOIObservationFilterSequence",
+    "hasObservation",
+    "hasAttribute",
+    "hasAdditional"
+]
+
+
+def handle_custom_functions(operator, args, existing_ggps: GroupGraphPatternSub | None = None, ):
+    ggps = existing_ggps if existing_ggps is not None else GroupGraphPatternSub()
+    focus_node_vot = VarOrTerm(varorterm=Var(value="focus_node"))
+    if operator in ["FOIObservationFilterDirect", "FOIObservationFilterSequence", "hasObservation"]:
+        obj = VarOrTerm(varorterm=Var(value="observation"))
+        foi_to_obs_tssp = create_tssp_alt_or_alt_inverse(
+            subject=focus_node_vot,
+            first_pred=IRI(value="http://www.w3.org/ns/sosa/isFeatureOfInterestOf"),
+            second_pred=IRI(value="http://www.w3.org/ns/sosa/hasFeatureOfInterest"),
+            obj=obj,
+            inverse_second_prop=True
+        )
+        if operator in ["FOIObservationFilterDirect", "FOIObservationFilterSequence"]:
+            filter_1_object = convert_value_to_rdf_term(args[1])
+            obs_filter_1 = TriplesSameSubjectPath.from_spo(
+                subject=obj.varorterm,
+                predicate=IRI(value=args[0]),
+                object=filter_1_object
+            )
+            filter_2_object = convert_value_to_rdf_term(args[3])
+            obs_filter_2 = None
+            if operator == "FOIObservationFilterDirect":
+                obs_filter_2 = TriplesSameSubjectPath.from_spo(
+                    subject=obj.varorterm,
+                    predicate=IRI(value=args[2]),
+                    object=filter_2_object
+                )
+            elif operator == "FOIObservationFilterSequence":
+                obs_filter_2 = create_tssp_sequence(
+                    subject=obj,
+                    pred_1=IRI(value=args[2]),
+                    pred_2=IRI(value=args[3]),
+                    obj=VarOrTerm(varorterm=GraphTerm(content=filter_2_object))
+                )
+            ggps.add_pattern(TriplesBlock.from_tssp_list([obs_filter_2, obs_filter_1, foi_to_obs_tssp]))
+        elif operator == "hasObservation":  # https://github.com/RDFLib/prez/issues/378#issuecomment-3269086241
+            """
+            {
+                ?focus sosa:isFeatureOfInterestOf/^sosa:hasFeatureOfInterest ?obs .
+                ?obs sosa:observedProperty $arg1 .
+                { ?obs sosa:hasSimpleResult $arg2 }
+                UNION
+                { ?obs sosa:hasResult $arg2 }
+                UNION
+                { ?obs sosa:hasResult/rdf:value $arg2}
+                VALUES $arg2 { "val1" }
+                VALUES $arg1 { <https://example.org/vocab/prop1> }
+            }"""
+            obs_prop_var = Var(value="obs_prop")
+            result_var = Var(value="result")
```

Comment:
Should these vars have a counter attached to the end? So that if you have multiple hasObservation conditions, or combination of hasObservation and hasAttribute, then the vars don't alias.

Reply:
Yes thank you I planned to do this and forgot - will add it in

Reply:
added in https://github.com/RDFLib/prez/pull/417/commits/b608ca76d545ae827ac869b0cbd26ddac0258749 example calls that would have overlapping variables added to the http example file